Improve gensim documentation (numfocus)

mpenkov opened the pull request:
The goal of the PR is:

>  The goal is to organize and clean up all the docs, aid users in information discovery. So they don't have to fumble around for "what's where": there should be a clear structure, with authoritative tutorials and how-tos in one place.

This PR refactors our existing documentation into a Sphinx Gallery. The PR offers several benefits:

- We no longer split our docs between RST and Jupyter notebooks, with no duplication. We write our docs as script.py, and Sphinx Gallery generates RST and Jupyter notebooks from them.
- The above script.py get executed each time we build documentation, so we can be sure that our examples still work.

The rendered results are available here: https://radimrehurek.com/gensim/gensim_numfocus/auto_examples/index.html

Fixes #2424. Fixes #1641.

piskvorky commented:
@gojomo can you have a look?

piskvorky commented:
**EDIT**: after reviewing these comments with @mpenkov now, all items from my previous review are fixed now except:

* "5: Site structure". We want to move back "About", "Support" and "Install" (with "Install" probably right on the main landing page; perhaps even with something like the install selection menu ala https://pytorch.org or spacy.io).
* "Downloader Module" is still too non-descript.
* Differentiation between tutorials and how-tos is much clearer now. I would even highlight in bold the salient phrases **learning-oriented lessons**  and **solve a specific problem**.

   But now that the section purpose is clear, some of the how-tos don't seem to fulfil it ("How to Perform Pivoted Document Length Normalization" solves no specific problem) => move them to Tutorial?
* Some clean up on the landing page: item margins in top menu, forgotten footer menu, installation section.

mpenkov commented on docs/src/Makefile:
```diff
@@ -41,7 +41,7 @@ html:
 
 upload:
-	scp -r _build/html/* rr:public_html/gensim/
+	scp -r _build/html/* rr:public_html/gensim/gensim_numfocus
```
Yes

gojomo commented:
> The problem with Jupyter notebooks is that they're difficult to maintain. Someone has to remember to to go through them "once in a while" and make sure they work, one by one. They're also hard to keep track of - if you look at our current notebooks subdirectory, its an unorganized mess.

I'd agree Jupyter Notebooks have some maintenance challenges, and perhaps aren't ideal for intro/canonical documentation. But my greatest interest is in there being a growing library of example notebooks, some intro-level but others focusing-in on more advanced techniques, that users encounter 'a la carte', when needed. For **that** use-case, which may be different from your current goals with this issue & other project priorities:

* contributor ease may be paramount – using a standard `.ipynb` as the unit-of-contribution rather  than a potentially unfamiliar tool like 'sphinx gallery' 
* while some logical organization is always beneficial, a somewhat messy "everything else" holding place may also be useful, to capture useful-but-not-quite-polished contributions
* it's potentially helpful for cells to throw example errors, or for some cells to show optional, highly-expensive operations – so an acceptance-test that involves running the whole notebook/script without error isn't appropriate (or practical on CI systems) – and simply having a notebook with cached output cells that "somebody ran intelligently, manually, recently" is good enough

So, even if mainline docs adopt a sphinx-inspired `script.py` format, I'd hate to the see the `notebooks` collection of other more-optional/more-advanced examples to either go away, or adopt a harder-to-contribute format. Maybe it becomes a `contrib` directory, or separate project with its own release practices, or project that's only ever cloned from github? 

(I can think of some standard practices that might help improve the maintainability of contributed notebooks. We could have some standard functions that must be run in the 1st cell, which report time/versions in effect when the output cells are cached. We might add some test/build time assertions that warn when notebooks get "too far" out of alignment, by date or version, with current releases. Etc - though perhaps this is a topic for another issue/forum.)

> > (Similarly, the links offering Download Python source code: run_word2vec.py or Download Jupyter notebook: run_word2vec.ipynb seem misguided: if these same docs/notebooks are already bundled inside gensim, anyone who'd be using them already has a local copy, and they should be pointed there – where they might find other notebooks, or where notebooks could potentially cross-reference each other via local paths, etc.)
> 
> People will only have access to those notebooks/scripts if they've cloned the gensim git repository. They won't be available if they're just run `pip install gensim`. I think it's worth keeping those links for the git-impaired.

I would lean toward including a copy of version-matched official docs, especially those that can be run as learning exercises, in the pip-installed package. It's not that much data, and being able to say "if you have gensim installed, you have these runnable examples as well" is pretty useful. (Alternately: make them available as a separate "gensim-docs" package that's still `pip`-installable and version-synchronized with the main package.)

> Yeah, on one hand, the api.load implementation is... suboptimal.
> 
> On the other hand, the use of `api.load` guarantees that the datasets we need for the tutorials are available. That's a big plus. We cannot expect the same of e.g. hyperlinks to external data. I think the way forward is to:
> 
> 1. Continue using api.load in documentation
> 2. Freeze the gensim-data project
> 3. Reimplement api.load to be better

I'm in favor of (3). We're not gaining much if anything by having small code-shims loaded & executed from an opaque URL, and it's bad security and sw-engineering practice. All executable code could easily be placed in a versioned project (whether `gensim` itself or a sibling), and only updated via that project's normal commit-and-release process. The supported datasets aren't changing/expanding that rapidly. If your gensim installation is going to run different code from when you last installed a named version, you should be explicitly installing an updated version!

> > While starting with loading an existing model, to show basic word-vec operations, is a good idea, using the multi-gigabyte GoogleNews set is asking for troulble. Anyone with <=4GB of RAM will probably error out, and maybe many with <=8GB RAM as well.
> 
> We could demonstrate the models on toy data, but then the results would be less meaningful (BTW, which tutorial are you referring to?)

The word2vec example, as its 1st tangible step, tried to load the `GoogleNews` data – and that quickly kills the kernel when run in the `MyBinder` service.

gojomo commented:
> Can you suggest a better way to organize? If the organization is not a net plus, we're not merging.

My main thoughts would be:

* be sure to group related algorithms, and then list tutorials in ascending-difficulty and descending-size-of-audience. Of the existing tutorials, those that are clearly in the "2Vec" family are: Word2Vec, Doc2Vec, FastText, WMD, Annoy. So those should be in a section that indicates they're related, and not interleaved with less-related tutorials (like 'Text Summarization' or 'LDA Model'). 
* the appearance & ordering of algorithms in official documentation will be read as expert recommendations, so listing an advanced optimization (like ANNOY) 1st will misguide many beginners. And including anything prominent about an algorithm of iffy API/quality (as 'Text Summarization', per the reasoning in #2592) is questionable. Docs should prioritize what's working best in the library. 
* include as much up-front text as possible to help explain/contextualize the tutorials/how-tos – "this is if you want to do X, Y, Z; skip if you need J, K, L." – in the overall list, not hidden behind mouse-over effects, so users have to do a minimum of clicking-into or mousing-over to scan for the most relevant units for their current task

mpenkov commented:
@piskvorky OK, I've finally got around to looking at this (the gensim bugfix release was taking up all my bandwidth). I think I've addressed all comments except for:

1. ordering of examples, as suggested by @gojomo 
2. Web server redirects
3. Move pivoted norm from guides to tutorials

I will take care of these next week.

Can you please double-check? Are there any other outstanding items?

mpenkov commented:
Here's a draft plan for rolling out these changes:

0. Delete old Jupyter notebooks that have already been ported to Sphinx Gallery
1. Merge this PR
2. Push documentation to new location (https://radimrehurek.com/gensim/ instead of https://radimrehurek.com/gensim/gensim_numfocus)
3. Set up server redirects
  - tutorial.html -> auto_examples/index.html#core-tutorials
  - intro.html -> auto_examples/core/run_core_concepts.html
  - install.html -> index.html#install
4. ...
5. PROFIT

@piskvorky Can you please have a look and let me know if I've missed anything?

gojomo commented:
Intro tutorials are still in an unhelpful order – with `Doc2Vec` (which should definitely go after `Word2Vec`, and perhaps after all other word2vec-centric topics). The details of what a lesson contains are still hidden behind a mouse-over. 

If others think this organization is more helpful for beginners than "native" Jupyter notebooks – I personally don't – I could see putting basic intro tutorials here. But I don't think more advanced/lengthy examples (like the "reproduce le/mikolov with Doc2Vec" & perhaps others) work well here. Some reasons:

* users at that level should be able to use real Jupyter notebooks
* such notebooks can take a lot of time and resources to run - for example, the Doc2Vec notebook requires 30 minutes plus & >4GB RAM - will that delay be added to every test-build? 
* some of the constructs here should only be encouraged among more-advanced users (more on this below)

I've only taken a close-look at the "reproduce le/mikolov with Doc2Vec" notebook, which seems a subset of the `doc2vec-imdb.ipynb` notebook I'd made. The work to integrate here seems to have added some polish, especially with regard to the narrative text. But it's also added some nonsense. Some thoughts:

* Unfortunately, no one has ever fully reproduced the Le/Mikolov "Paragraph Vector" results – and some statements from Mikolov suggest he doesn't know how Le got the results reported in the paper. This notebook just approximates their approach/result on one of their multiple "experiments". So the title is overpromising a bit.
* It seems to intend to "shrink" the corpus to run faster - but that line is currently commented out. What's the intended final policy? (I'm not sure other parts of the notebook will work or give meaningful results with an arbitrarily-shrunk corpus.)
* The part "Sanity checking. Let’s see if our models give meaningful results" is nonsense, because at that point the models haven't been trained at all - the results aren't meaningful, but the narrative just skips that without comment. 
* Given that this version of the notebook doesn't use a single shared `build_vocab()` (with hand-tampering to move the results to all models), and doesn't do per-epoch evaluations – which are both advanced, error-prone techniques that I agree shouldn't be in an introductory notebook – the 'build_vocab()` and 'train()` steps shouldn't be so far apart - given they are more-typically done one-after another. 
* Though the top of the tutorial shows logging set at INFO level, there's no logging output evident in executed cells. Is this a side-effect of the Sphinx processing? (I find it confusing.)

If this were a plain notebook in the repo, I'd be more likely to jump in and fix these sorts of issues. In the Sphinx format, because I'm unfamiliar with that tool, I'm unlikely to fix myself. 

Also, though I'd agree certain advanced optimizations, & optional/"slop" cells, shouldn't (and can't!) go into these Sphinx-format docs, I think it'd be useful to offer those somewhere. If the plain `doc2vec-imdb.ipynb` notebook is deleted to reduce duplication, they're lost. If that notebook and this are retained, there's duplication likely gradual divergence as fixes are applied one place but not the other. 

I guess my point is: if we're keeping any 'advanced topics' notebooks separate from the Sphinx-ified material, I'd keep this notebook separate. We could just point to it at the end of the 'basic' Doc2Vec notebook, as something that can be run in a different format. 

I haven't taken a close look at the other 'how-to' lessons, but a quick glance at "How to Classify News Articles into Topics" shows *no* actual supervised classification occurring – only unsupervised topic-modeling. So that again seems like lesson-title likely to misguide those looking for true classification.

mpenkov commented:
Thank you for the feedback @gojomo . Please see my responses inline below.

> Intro tutorials are still in an unhelpful order – with `Doc2Vec` (which should definitely go after `Word2Vec`, and perhaps after all other word2vec-centric topics).

Changing the order of tutorials is fairly trivial. If you have an optimal order, we can apply it [here](https://github.com/mpenkov/gensim/blob/4d85ffd069d33791f02ff62b67e5801e26c36a1c/docs/src/conf.py#L237-L263).

> The details of what a lesson contains are still hidden behind a mouse-over.

Yes, moving lesson descriptions out of the mouse-overs may be helpful, but it isn't something that Sphinx Gallery supports at the moment.

> If others think this organization is more helpful for beginners than "native" Jupyter notebooks – I personally don't – I could see putting basic intro tutorials here. But I don't think more advanced/lengthy examples (like the "reproduce le/mikolov with Doc2Vec" & perhaps others) work well here. Some reasons:
> 
>     * users at that level should be able to use real Jupyter notebooks

Jupyter notebooks are still available for people who want to use them (Sphinx Gallery converts the Py scripts to notebooks [here](https://github.com/mpenkov/gensim/tree/numfocus/docs/src/auto_examples)). So, the users haven't lost anything at this stage.
> 
>     * such notebooks can take a lot of time and resources to run - for example, the Doc2Vec notebook requires 30 minutes plus & >4GB RAM - will that delay be added to every test-build?
> 

No. We don't run the gallery documentation on each build, for the reasons you described.

Please see [this README.rst](https://github.com/mpenkov/gensim/blob/numfocus/docs/src/gallery/README.rst) for a description of what goes into version control and why.

>     * some of the constructs here should only be encouraged among more-advanced users (more on this below)

We can put those in a separate section explicitly labeled "Advanced".

> 
> I've only taken a close-look at the "reproduce le/mikolov with Doc2Vec" notebook, which seems a subset of the `doc2vec-imdb.ipynb` notebook I'd made. The work to integrate here seems to have added some polish, especially with regard to the narrative text. But it's also added some nonsense. Some thoughts:
> 
>     * Unfortunately, no one has ever fully reproduced the Le/Mikolov "Paragraph Vector" results – and some statements from Mikolov suggest he doesn't know how Le got the results reported in the paper. This notebook just approximates their approach/result on one of their multiple "experiments". So the title is overpromising a bit.
>

Yeah, I had a lot of trouble deciding what section of the documentation that should go into, and what the title for it should be. It's really neither of:

1. A learning-oriented tutorial designed to show beginners how to use the system
2. A goal-oriented guide designed to show people how to use gensim to solve a specific real-world problem

This is partially because the motivation behind that Jupyter notebook isn't 100% clear. In the end, I went with 2), where I defined the real-world problem as replicating the results of the paper.

>     * It seems to intend to "shrink" the corpus to run faster - but that line is currently commented out. What's the intended final policy? (I'm not sure other parts of the notebook will work or give meaningful results with an arbitrarily-shrunk corpus.)
> 

Yeah, that's cruft. We can get rid of that. I'll push a commit.

>     * The part "Sanity checking. Let’s see if our models give meaningful results" is nonsense, because at that point the models haven't been trained at all - the results aren't meaningful, but the narrative just skips that without comment.
>

Is that something that got added as part of this PR? Or was it present in the original as well?

>     * Given that this version of the notebook doesn't use a single shared `build_vocab()` (with hand-tampering to move the results to all models), and doesn't do per-epoch evaluations – which are both advanced, error-prone techniques that I agree shouldn't be in an introductory notebook – the 'build_vocab()` and 'train()` steps shouldn't be so far apart - given they are more-typically done one-after another.
> 
>     * Though the top of the tutorial shows logging set at INFO level, there's no logging output evident in executed cells. Is this a side-effect of the Sphinx processing? (I find it confusing.)

Yes, I don't think Sphinx includes the logging output in the HTML.

> 
> If this were a plain notebook in the repo, I'd be more likely to jump in and fix these sorts of issues. In the Sphinx format, because I'm unfamiliar with that tool, I'm unlikely to fix myself.
>

I think if you look at the [source](https://github.com/mpenkov/gensim/blob/numfocus/docs/src/auto_examples/howtos/run_doc2vec_imdb.py) of that tutorial, you'll see that editing it isn't significantly different to what you'd do to a Jupyter notebook. Yes, there's an additional burden to learning how to get things going with Sphinx gallery, but I've tried to mitigate that with additional documentation:

- [How to Author Documentation](https://radimrehurek.com/gensim/gensim_numfocus/auto_examples/howtos/run_doc.html): this guide aims to help prospective doco authors
- [README.rst](https://github.com/mpenkov/gensim/blob/numfocus/docs/src/gallery/README.rst): describes how the Gallery works

On top of that, the Gallery offers significant benefits over Jupyter notebooks. Some of these we may have discussed earlier:

1. It's easy to confirm the documentation still works. You run the Python script through the interpreter, and if it doesn't crash, then success. With Jupyter, you have to fire up the notebook server, load up the notebook, and only then run it.
2. It's easy to do version control. It's a plain Python script. You can use regular git features like blame and diff. With Jupyter, you can't do any of these things.
3. It's easy to get an automatically-generated top-level view of the gallery showing all examples. Yes, the mouse-overs are a point of contention, but it's still more helpful than showing people a directory listing of half-working Jupyter notebooks, or manually maintaining listings like [this](https://github.com/mpenkov/gensim/blob/numfocus/tutorials.md).

> Also, though I'd agree certain advanced optimizations, & optional/"slop" cells, shouldn't (and can't!) go into these Sphinx-format docs, I think it'd be useful to offer those somewhere. If the plain `doc2vec-imdb.ipynb` notebook is deleted to reduce duplication, they're lost. If that notebook and this are retained, there's duplication likely gradual divergence as fixes are applied one place but not the other.
> 

As you point out, we definitely don't want multiple editable copies of the same documentation. If we _really_ want to store slop, we can do it via other means:

- Comment it out
- Put it in a function that does not get called by default

> I guess my point is: if we're keeping any 'advanced topics' notebooks separate from the Sphinx-ified material, I'd keep this notebook separate. We could just point to it at the end of the 'basic' Doc2Vec notebook, as something that can be run in a different format.
>

We can remove the notebook from the gallery, but I think discovery will suffer. How will people find it? I think it's better to keep it in the Gallery, but label it as an advanced example and make the caveats obvious (don't try this at home, kids).

> 
> I haven't taken a close look at the other 'how-to' lessons, but a quick glance at "How to Classify News Articles into Topics" shows _no_ actual supervised classification occurring – only unsupervised topic-modeling. So that again seems like lesson-title likely to misguide those looking for true classification.

OK, in that case it's a poor title. What would be a better title, in your opinion?

gojomo commented:
> Changing the order of tutorials is fairly trivial. If you have an optimal order, we can apply it [here](https://github.com/mpenkov/gensim/blob/4d85ffd069d33791f02ff62b67e5801e26c36a1c/docs/src/conf.py#L237-L263).

I made a [previous suggestion](https://github.com/RaRe-Technologies/gensim/pull/2591#issuecomment-531978680) for the ordering of the topics with which I'm most familiar. 

> > * users at that level should be able to use real Jupyter notebooks
> > ```
> 
> Jupyter notebooks are still available for people who want to use them (Sphinx Gallery converts the Py scripts to notebooks [here](https://github.com/mpenkov/gensim/tree/numfocus/docs/src/auto_examples)). So, the users haven't lost anything at this stage.

But, if the main target audience for a certain topic is a proper notebook, run as a notebook – and if further for other advanced reasons a true notebook is **better**  (like the desire to demonstrate errors or optional cells inline, or include processes that may take hours to run, or the desire to include text that comments on the mildly-random output of a particular run) – then the extra non-notebook format (`.py`) & presentation (`.rst`) just means extra steps and limitations for the maintainer, and duplication for the reader. 

> > * Unfortunately, no one has ever fully reproduced the Le/Mikolov "Paragraph Vector" results – and some statements from Mikolov suggest he doesn't know how Le got the results reported in the paper. This notebook just approximates their approach/result on one of their multiple "experiments". So the title is overpromising a bit.
> 
> Yeah, I had a lot of trouble deciding what section of the documentation that should go into, and what the title for it should be. It's really neither of:
> 
> 1. A learning-oriented tutorial designed to show beginners how to use the system
> 2. A goal-oriented guide designed to show people how to use gensim to solve a specific real-world problem
> 
> This is partially because the motivation behind that Jupyter notebook isn't 100% clear. In the end, I went with 2), where I defined the real-world problem as replicating the results of the paper.

A more descriptive title is fine – and ideally, then, it would include Le/Mikolov's term "Paragraph Vector". But, it should avoid raising hopes that this is a full or successful reproduction. Perhaps, "Le/Mikolov 'Paragraph Vector' IMDB experiment with gensim Doc2Vec"?

> > * The part "Sanity checking. Let’s see if our models give meaningful results" is nonsense, because at that point the models haven't been trained at all - the results aren't meaningful, but the narrative just skips that without comment.
> > ```
> 
> Is that something that got added as part of this PR? Or was it present in the original as well?

It's not in the original <https://github.com/RaRe-Technologies/gensim/blob/develop/docs/notebooks/doc2vec-IMDB.ipynb>, so I'm not sure who added it when.

> Yes, I don't think Sphinx includes the logging output in the HTML.

It seems suboptimal that users who run exactly this code will thus see something different than the Sphinx rendering shows. (But then, removing that cell would guide users away from a useful practice.) Sphinx's tradeoffs are not what I'd personally choose.

> > If this were a plain notebook in the repo, I'd be more likely to jump in and fix these sorts of issues. In the Sphinx format, because I'm unfamiliar with that tool, I'm unlikely to fix myself.
> 
> I think if you look at the [source](https://github.com/mpenkov/gensim/blob/numfocus/docs/src/auto_examples/howtos/run_doc2vec_imdb.py) of that tutorial, you'll see that editing it isn't significantly different to what you'd do to a Jupyter notebook. Yes, there's an additional burden to learning how to get things going with Sphinx gallery, but I've tried to mitigate that with additional documentation:

My usual process of authoring such examples involves interactively using a real notebook: trying different variants of cells, running cells out-of-order, and so forth. Sure, I'll run it top-to-bottom, once, in order at the end – but that's a final checking step. 

There's essentially zero chance I'd want to initially author (or interactively update/debug!) in that special-`.py`-format-that-becomes-a-notebook-later. The other way around (notebook-to-sphinx's-`.py`) would be far more important to me, so I could author/update as a notebook, then convert. (I assume there's some Sphinx tool for that?) And yet, having to keep running that would still be a mildly inconvenient extra step, especially when making those repeated "one last thing" changes that come up in final reviews.

> On top of that, the Gallery offers significant benefits over Jupyter notebooks. Some of these we may have discussed earlier:
> 
> 1. It's easy to confirm the documentation still works. You run the Python script through the interpreter, and if it doesn't crash, then success. With Jupyter, you have to fire up the notebook server, load up the notebook, and only then run it.

For some basic straight-through notebooks, yes, this is very helpful. 

But for other notebooks with inline demos of errors, or mutually-exclusive or costly optional paths, this is constraining – essentially making those actions impossible.  So I'd still want there to be a place for 'advanced' notebooks that don't have the overhead, and constraints, of the Sphinx Gallery approach. 

(I also suspect a few other lighter-weight conventions could help make native notebooks more maintainable. I mentioned [a couple ideas for that in this above](https://github.com/RaRe-Technologies/gensim/pull/2591#issuecomment-531975991), and I'm pretty sure that at least experimentally, we've previously had auto-build-test setups that could optionally run notebooks to check for errors.)

> 2. It's easy to do version control. It's a plain Python script. You can use regular git features like blame and diff. With Jupyter, you can't do any of these things.

Agreed, the fact that diffs in `.ipynb` files are really ugly/unhelpful is, from my perspective, the strongest argument for the `.py` format. 

But it also seems that benefit would be obtainable, while retaining the flexibility/ease of `.ipynb` being the canonical/authoring format, by auto-generating paired-`.py` files for diff purposes. 

> 3. It's easy to get an automatically-generated top-level view of the gallery showing all examples. Yes, the mouse-overs are a point of contention, but it's still more helpful than showing people a directory listing of half-working Jupyter notebooks, or manually maintaining listings like [this](https://github.com/mpenkov/gensim/blob/numfocus/tutorials.md).

Again, that's helpful for a strongly-curated & well-ordered set of interconnected intro lessons, but much less important for an 'a la carte' grab-bag of advanced & miscellaneous topics. (Also, a convention of some required boilerplate fields atop contributed notebooks could make an autogenerated listing as easy as with Sphinx.)

> As you point out, we definitely don't want multiple editable copies of the same documentation. If we _really_ want to store slop, we can do it via other means:
> 
> * Comment it out
> * Put it in a function that does not get called by default

Yes, those could be workarounds, but may result in more awkward presentations for the reader. For example, the attempt to demonstrate a `KeyError` – but ultimately showing nothing  – as in [this section of the word2vec lesson](https://radimrehurek.com/gensim/gensim_numfocus/auto_examples/010_tutorials/run_word2vec.html#word2vec-demo), is going to be really opaque to people who just barely understand exceptions.

> > I guess my point is: if we're keeping any 'advanced topics' notebooks separate from the Sphinx-ified material, I'd keep this notebook separate. We could just point to it at the end of the 'basic' Doc2Vec notebook, as something that can be run in a different format.
> 
> We can remove the notebook from the gallery, but I think discovery will suffer. How will people find it? I think it's better to keep it in the Gallery, but label it as an advanced example and make the caveats obvious (don't try this at home, kids).

Well, there are many dozens of places I've linked to this notebook's prior location (<https://github.com/RaRe-Technologies/gensim/blob/develop/docs/notebooks/doc2vec-IMDB.ipynb>), and it could also be linked at the end of the "intro" notebook. 

But I'm not just concerned about this notebook, but rather all rough and/or advanced and/or peculiar notebooks-of-its-kind. Subsuming such contributions into the gallery format places extra barriers of new steps, formats, and content-limitations on authors. There should be a place for such marginal stuff, and having that place be disorganized is better than adding new barriers to contribution/maintenance.
 
> > I haven't taken a close look at the other 'how-to' lessons, but a quick glance at "How to Classify News Articles into Topics" shows _no_ actual supervised classification occurring – only unsupervised topic-modeling. So that again seems like lesson-title likely to misguide those looking for true classification.
> 
> OK, in that case it's a poor title. What would be a better title, in your opinion?

Perhaps, "Unsupervised Topic Modeling of News Articles"? (But: best if the material's author chooses an appropriate title.)

Also: I think all my concerns about particulars of the word2vec lesson from [this above comment](https://github.com/RaRe-Technologies/gensim/pull/2591#issuecomment-531032247) remain, though perhaps you want this to be a "reorg" PR and only address content issues later.

piskvorky commented:
We reviewed all comments from Misha, cleaned up the structure, language, dependencies etc. Where we're at now:

- The new site and workflow is good to go.
  - New workflow = vetted tutorials and how-tos stored in **`.py` format as the source of truth**, with `.rst` and `.ipynb` generated from that.
  - I tried editing some `.py` tutorials today to test the workflow, and it's easy enough. The `.py` with rst in comments is easier to track and diff (and IMO also to author and edit, but YMMV) than Jupyter.
  - In case we receive an external contribution that is worthwhile to be included under tutorials/how-tos in a *different* format (e.g. a Jupyter notebook), we'll adapt it to `.py` manually. I expect this will be vanishingly rare: pretty much just @gojomo.
- Added a new "Other resources" section under the Documentation page.
  - "Other resources" serves to aid discovery of both non-standard resources (videos, blogs, etc), as well as notebooks and generally things that don't fit into our standard workflow.
  - Lower standard on quality and repeatability than standard tutorials and how-tos.
  - We want to minimize the amount of resources here, only keep the really strong / useful ones.
  - Some old resources, such as "Experiments on the English Wikipedia", will be cleaned up + "upgraded" to the proper tutorial format in time, one by one. Not blocking for this PR.
- @mpenkov compiled a list of all files removed in this PR along with a link to their new, reformatted / cleaned-up version. These were ~18 notebooks and documents.
  - No files were removed without a replacement. Some "old" notebooks are still preserved under `/notebooks` in the github repo, even though nothing links to them any more = they are not discoverable. We'll be looking to clean up and transfer what's good into the new workflow, eventually, gradually.
  - This list will be used to set up web server redirects once the new site goes live, to make sure that existing links around the internet continue to work.

TODO:
- [ ] Clean up the "Other resources" gallery section.
- [ ] Improve the tutorial images: bias the word clouds, add missing images.
- [ ] Test the "new" website end-to-end locally.
- [ ] Test website online under the temp location at https://radimrehurek.com/gensim/gensim_numfocus/
- [ ] Merge PR and push website "live" at https://radimrehurek.com/gensim/
- [ ] Enable the web server redirects for removed resources (see above)

piskvorky commented:
@mpenkov after merging this PR, Github showed me this alert:
<img width="1224" alt="Screen Shot 2019-10-22 at 01 03 24" src="https://user-images.githubusercontent.com/610412/67249453-04740800-f468-11e9-8bf7-94e91709d468.png">

What is that `requirements.txt` file for?

gojomo commented:
Dozens (maybe hundreds) of links from elsewhere to places like <https://github.com/RaRe-Technologies/gensim/blob/develop/docs/notebooks/doc2vec-IMDB.ipynb> now give an unhelpful Github "not-found" page; placing tombstone docs at the old locations pointing to the canonical new locations would help people find the new formats.

mhaeussermann commented:
> We'll also need to revise the README on the main Github page – its [Documentation links](https://github.com/RaRe-Technologies/gensim#documentation) don't work any more.

Came here because I noticed this.

piskvorky commented:
Agreed. @mpenkov if other documents were removed (not covered by the redirects on our website), let's find a way to "redirect" them too. Since we cannot redirect github locations, maybe a placeholder document? 404-not-found is not good.

piskvorky commented:
I'm -1 on keeping a top-level `requirements.txt` that does not contain actual requirements. It's confusing (besides triggering that Github warning).

Regarding the Github redirects: how about a simple single document that shows "Documentation moved. See https://radimrehurek.com/gensim/auto_examples/index.html for up-to-date Gensim tutorials and notebooks."

piskvorky commented:
Btw is the URL `/auto_examples` auto-generated? `/documentation` would be marginally better (nothing critical).

mpenkov commented:
Yeah, we can do that. I'll take care of it in a separate PR once the more immediate issues are solved.

piskvorky commented:
Great, thanks @mpenkov. I'll update the redirects too, when that happens.
